Fix bazel dependency failure for non-standard cuda toolkit path

nsuke opened the pull request:
ArchLinux cuda package puts the toolkit to `/opt/cuda` without symlinks (no /opt/cuda-7.5 etc).

tensorflow-jenkins commented:
Can one of the admins verify this patch?

davidzchen commented on third_party/gpus/cuda_configure.bzl:
```diff
@@ -403,12 +403,16 @@ def _create_cuda_repository(repository_ctx):
        })
 
   # Set up crosstool/
+  cuda_include_path = '%s-%s/include' % (cuda_toolkit_path, cuda_version)
```
Can you add a comment describing the non-standard CUDA toolkit path that makes this necessary?

Also, it would be better to move this into a helper function, similar to `_find_cudnn_header_dir()`.

davidzchen commented:
Jenkins, test this please.

nsuke commented:
@davidzchen addressed your comment.

Just noticed that you were already planning to fix it (#4096).
That would be diffinitely better when bazel requirement is met.

davidzchen commented on third_party/gpus/cuda_configure.bzl:
```diff
@@ -403,12 +427,14 @@ def _create_cuda_repository(repository_ctx):
        })
 
   # Set up crosstool/
+  cuda_include_path = _real_cuda_toolkit_path() + '/include'
```
Why is this function called with no parameters?

nsuke replied:
Wow. Fixed and ran script once.

davidzchen commented on third_party/gpus/cuda_configure.bzl:
```diff
@@ -403,12 +427,15 @@ def _create_cuda_repository(repository_ctx):
        })
 
   # Set up crosstool/
+  cuda_include_path = _real_cuda_toolkit_path(repository_ctx, cuda_toolkit_path,
```
If you are having this function return the CUDA toolkit path, perhaps combine it with `_cuda_toolkit_path()`?

nsuke replied:
Makes sense. Done.

davidzchen commented on third_party/gpus/cuda_configure.bzl:
```diff
@@ -237,6 +237,30 @@ def _check_dir(repository_ctx, directory):
+  # Handle typical configuration where real path is <basedir>/cuda-<version>
+  # and the provided path is <basedir>/cuda.
+  version_suffixed = "%s-%s" % (cuda_toolkit_path, cuda_version)
+  if repository_ctx.path(version_suffixed).exists:
```
Also check that `cuda_version` is non-empty here?

nsuke replied:
Done.

davidzchen commented:
Jenkins, test this please.

davidzchen commented:
Jenkins, test this please.

davidzchen commented on third_party/gpus/cuda_configure.bzl:
```diff
@@ -120,13 +120,33 @@ def _enable_cuda(repository_ctx):
+def _cuda_toolkit_path(repository_ctx, cuda_version):
+  """Finds the cuda toolkit directory.
+
+   If cuda_version is given, attempts to return a real path.
```
Seems slightly redundant since it still attempts to find a real path even if cuda_version is not given.

davidzchen commented on third_party/gpus/cuda_configure.bzl:
```diff
@@ -120,13 +120,33 @@ def _enable_cuda(repository_ctx):
+    version_suffixed = "%s-%s" % (cuda_toolkit_path, cuda_version)
+    if repository_ctx.path(version_suffixed).exists:
+      return version_suffixed
+  # Now hoping that the provided path was already a real path.
```
Returns the non-versioned path if cuda version is not provided or if the installation does not use a cuda-<version> directory, such as on ArchLinux where CUDA installs directly to /opt/cuda.

davidzchen commented:
Thanks! LGTM aside from a couple of minor comments.

nsuke committed:
Fix bazel dependency failure for non-standard cuda toolkit path

Build script was putting partially hardcoded path to cxx_builtin_include_directory.

nsuke commented:
@davidzchen tweaked the comments as per your suggestions.
Looks much nicer, thanks :)

davidzchen commented:
Jenkins, test this please.

davidzchen commented:
@vrv This change looks good to me. Can you take a look and merge if there is no other feedback? Thanks!
